core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessage.java

```diff
@@ -16,276 +16,58 @@
-        return completionFuture;
-    }
-
-    /**
-     * Callback invoked by {@link Subscription#request(long)} to add {@code n} to demand.
-     */
-    abstract void request(long n);
-
-    /**
-     * Callback invoked by {@link Subscription#cancel()} to cancel the stream.
-     */
-    abstract void cancel();
-
-    /**
-     * Invoked after an element is removed from the {@link StreamMessage} and before
-     * {@link Subscriber#onNext(Object)} is invoked.
-     *
-     * @param obj the removed element
-     */
-    protected void onRemoval(T obj) {}
-
-    static void failLateSubscriber(SubscriptionImpl actualSubscription, SubscriptionImpl lateSubscription) {
-        final Subscriber<?> actualSubscriber = actualSubscription.subscriber();
-        final Subscriber<?> lateSubscriber = lateSubscription.subscriber();
-        final Throwable cause = abortedOrLate(actualSubscriber);
-
-        if (lateSubscription.needsDirectInvocation()) {
-            handleLateSubscriber(lateSubscriber, cause);
-        } else {
-            lateSubscription.executor().execute(() -> {
-                handleLateSubscriber(lateSubscriber, cause);
-            });
+    public <U> CompletableFuture<U> aggregate(AggregationOptions<T, U> options) {
+        final ByteBufAllocator alloc = options.alloc();
+        final SubscriptionOption[] subscriptionOptions = alloc != null ? POOLED_OBJECTS : EMPTY_OPTIONS;
+        if (!options.cacheResult()) {
+            return collect(options.executor(), subscriptionOptions).thenApply(options.aggregator());
         }
-    }
 
-    private static void handleLateSubscriber(Subscriber<?> lateSubscriber, Throwable cause) {
-        try {
-            lateSubscriber.onSubscribe(NoopSubscription.get());
-            lateSubscriber.onError(cause);
-        } catch (Throwable t) {
-            throwIfFatal(t);
-            logger.warn("Subscriber should not throw an exception. subscriber: {}", lateSubscriber, t);
+        final CompletableFuture<?> aggregation = this.aggregation;
+        if (aggregation != null) {
+            //noinspection unchecked
+            return (CompletableFuture<U>) aggregation;
         }
-    }
-
-    final T prepareObjectForNotification(T o, boolean withPooledObjects) {
-        onRemoval(o);
-        return touchOrCopyAndClose(o, withPooledObjects);
-    }
 
-    /**
-     * Returns newly created {@link CloseEvent} if the specified {@link Throwable} is not an instance of
-     * {@link CancelledSubscriptionException#INSTANCE} or {@link AbortedStreamException#INSTANCE}.
-     */
-    static CloseEvent newCloseEvent(Throwable cause) {
-        if (cause == CancelledSubscriptionException.INSTANCE) {
-            return CANCELLED_CLOSE;
-        } else if (cause == AbortedStreamException.INSTANCE) {
-            return ABORTED_CLOSE;
+        final CompletableFuture<U> future =
+                collect(options.executor(), subscriptionOptions).thenApply(options.aggregator());
```

Comment:
Question) Isn't it still possible that multiple invocations to `collect` occur?

e.g.
```
// Thread1
streamMessage.aggregate(options).join();
// Thread2
streamMessage.aggregate(options).join();
// Both threads reach this line, which results in multiple subscriptions on a single StreamMessage
```

Reply:
The request for `collect()` of the latter will be ignored and piggybacked on the first one.
https://github.com/line/armeria/blob/7182181955b78ffd1039a76ee859a25567defba6/core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessage.java#L70

Reply:
That's a good point and a nice reproducible test case.
I think we can probably solve this by setting a future to the `aggregationUpdater` before calling `collect` method.
If the future is set the `StreamMessage` becomes not subscribable anymore.
If the `StreamMessage` is subscribed by another that does not want to cache the value, it would set `aggregationUpdater` with a `Object null = new Object()` value.

Reply:
I understood the problem and thanks for the reproducible example. Let me handle it.